File: src/compiler/interpreter/SymbolTable.java
package compiler.interpreter;

import compiler.lexer.token.Token;
import compiler.lexer.token.TypeToken;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public class SymbolTable {
  private Map<String, Object> symtab = new HashMap<>();
  private Map<String, TypeToken> types = new HashMap<>();

  public TypeToken getSymbolType(Token token) {
    final var type = this.types.get(this.hashcode(token));

    if (type == null) {
      throw new UndeclaredIdentifierException(token);
    }

    return type;
  }

  public void setSymbolType(Token token, TypeToken type) {
    final var oldType = this.types.put(this.hashcode(token), type);

    if (oldType != null) {
      throw new RuntimeException(token.getValue() + "'s datatype cannot be redefined");
    }
  }

  public void setSymbolValue(Token token, Object value) {
    final var previous = this.symtab.put(hashcode(token), value);
  }

  public boolean hasSymbol(Token token) {
    return this.symtab.containsKey(hashcode(token));
  }

  public Object getValue(Token token) {
    return this.symtab.get(hashcode(token));
  }

  public String getAddress(Token token) {
    return this.hashcode(token);
  }

  public Object getValueAtAddress(String address) {
    return this.symtab.get(address);
  }

  public Object removeSymbol(Token token) {
    return this.symtab.remove(hashcode(token));
  }

  private String hashcode(Token token) {
    return token.getValue() + ":" + token.getClass().getSimpleName();
  }

  @Override
  public String toString() {
    return symtab.entrySet()
      .stream()
      .map(entry -> entry.getKey() + ":" + this.types.get(entry.getKey()).getValue() + " = " + entry.getValue().toString())
      .collect(Collectors.joining("\n"));
  }
}

File: src/compiler/interpreter/SymbolTable.kt
package compiler.interpreter

import compiler.parser.Symbol.Terminal
import compiler.parser.Symbol.Terminal.Keyword.Type
import java.util.*
import java.util.stream.Collectors

class SymbolTable {
    private val symtab: MutableMap<String, Any?> = HashMap()
    private val types: MutableMap<String, Type> = HashMap()

    fun getSymbolType(terminal: Terminal): Type {
        return types[hashcode(terminal)] ?: throw UndeclaredIdentifierException(terminal)
    }

    fun setSymbolType(terminal: Terminal, type: Type) {
        val oldType = types.put(hashcode(terminal), type)
        if (oldType != null) {
            throw RuntimeException(terminal.str + "'s datatype cannot be redefined")
        }
    }

    fun setSymbolValue(terminal: Terminal, value: Any?) {
        symtab[hashcode(terminal)] = value
    }

    fun hasSymbol(terminal: Terminal): Boolean {
        return symtab.containsKey(hashcode(terminal))
    }

    fun getValue(terminal: Terminal): Any {
        return symtab[hashcode(terminal)] ?: throw UndeclaredIdentifierException(terminal)
    }

    fun getAddress(terminal: Terminal): String {
        return hashcode(terminal)
    }

    fun getValueAtAddress(address: String?): Any? {
        return symtab[address]
    }

    fun removeSymbol(terminal: Terminal): Any? {
        return symtab.remove(hashcode(terminal))
    }

    private fun hashcode(terminal: Terminal): String {
        return terminal.str + ":" + terminal.javaClass.simpleName
    }

    override fun toString(): String {
        return symtab.entries
                .stream()
                .map {
                    it.key + ":" + types[it.key]!!.str + " = " + it.value.toString()
                }
                .collect(Collectors.joining("\n"))
    }
}
